src/jwe.c

```diff
@@ -794,6 +795,22 @@ static int _r_concat_kdf(jwe_t * jwe, jwa_alg alg, const gnutls_datum_t * Z, gnu
     kdf->data[kdf->size+3] = (unsigned char)(key_data_len) & 0xFF;
     kdf->size += 4;
 
+    size_t derived_key_len = _r_get_key_size(jwe->enc);
+    size_t current_key_len = 0;
+
+    for (uint8_t i = 1; ; i++) {
```

Comment:
If I follow-up the code and the documentation, the loop would have at least 2 steps max, so I uggest to replace the for loop with 2 `gnutls_hash_fast` where the second would be only if `derived_key_len > gnutls_hash_get_len(GNUTLS_DIG_SHA256)`, also for a better readability of the code.

Something like that:

```C
    memset(kdf->data+3, 1, 1);
    if (gnutls_hash_fast(GNUTLS_DIG_SHA256, kdf->data, kdf->size, derived_key) != GNUTLS_E_SUCCESS) {
      y_log_message(Y_LOG_LEVEL_ERROR, "_r_concat_kdf - Error gnutls_hash_fast (1)");
      ret = RHN_ERROR;
      break;
    }
    if (alg == R_JWA_ALG_ECDH_ES && derived_key_len > gnutls_hash_get_len(GNUTLS_DIG_SHA256)) {
      memset(kdf->data+3, 2, 1);
      if (gnutls_hash_fast(GNUTLS_DIG_SHA256, kdf->data, kdf->size, derived_key) != GNUTLS_E_SUCCESS) {
        y_log_message(Y_LOG_LEVEL_ERROR, "_r_concat_kdf - Error gnutls_hash_fast (2)");
        ret = RHN_ERROR;
        break;
      }
    }
```

What do you think?

Reply:
this is not required. 
1. you are suggesting loop-unrolling (even if 2 steps) and 
2. the kdf->data was incorrect in the 2nd step. I see that you can since updated the comment.
no potential benefit to this.

Reply:
I agree my method isn't required for the code itself, but it makes the process more readable. And someone else, or my future self will more likely understand the intention without having to know the RFC by heart.

Also, even if it doesn't change the expected behavior, the `break` statement on line 805 will end the `for` loop, but not the `do while(0);` loop.
This doesn't have any effect in here, because the `while(0)` is right below, but if some code is added afterwards, this may lead to errors.
A workaround is to add an `if (ret != RHN_OK) {break;}` statement right after the `for` loop.